zcash_client_sqlite/src/wallet.rs

```diff
@@ -519,28 +521,264 @@ pub(crate) fn get_balance(
-        INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
-        WHERE account = ? AND spent IS NULL AND transactions.block <= ?",
-        [u32::from(account), u32::from(anchor_height)],
-        |row| row.get(0).or(Ok(0)),
+    fully_scanned_height: BlockHeight,
+    chain_tip_height: BlockHeight,
+) -> Result<Option<Ratio<u64>>, SqliteClientError> {
+    // Get the number of notes scanned in the incomplete range directly from the blocks table
+    let scanned = conn
+        .query_row(
+            "SELECT SUM(sapling_output_count)
+             FROM blocks
+             WHERE height >= :fully_scanned_height",
+            named_params![":fully_scanned_height": u32::from(fully_scanned_height)],
+            |row| row.get::<_, Option<u64>>(0),
+        )
+        .optional()?
+        .flatten()
+        .unwrap_or(0);
+
+    if fully_scanned_height == chain_tip_height && scanned > 0 {
+        Ok(Some(Ratio::new(scanned, scanned)))
+    } else {
+        // We don't have complete information on how many outputs will exist in the shard at
+        // the chain tip without having scanned the chain tip block, so instead we estimate
+        // by linear regression.
+        let stmt_final_shards = conn.prepare(
+            "SELECT 
+                 shard_index, 
+                 start_position,
+                 subtree_start_height, 
+                 subtree_end_height
+             FROM v_sapling_shard_scan_state
+             ORDER BY shard_index DESC
+             LIMIT 2",
+        )?;
+
+        struct ShardRow {
+            shard_index: u32,
+            start_position: u64,
+            subtree_start_height: u32,
+            subtree_end_height: Option<u32>,
+        }
+
+        let read_row = |row: &rusqlite::Row| -> Result<ShardRow, SqliteClientError> {
+            Ok(ShardRow {
+                shard_index: row.get::<_, u32>(0)?,
+                start_position: row.get::<_, u64>(1)?,
+                subtree_start_height: row.get::<_, u32>(2)?,
+                subtree_end_height: row.get::<_, Option<u32>>(3)?,
+            })
+        };
+
+        let rows = stmt_final_shards.query([])?;
+        let tip_row = rows.next()?.map(read_row).transpose()?;
+        let prev_row = rows.next()?.map(read_row).transpose()?;
+
+        if let Some(tip) = tip_row {
+            let tip_delta = u32::from(chain_tip_height) - tip.subtree_start_height;
+
+            if let Some(prev) = prev_row {
+                // use the previous row to interpolate the recent rate of output creation
+                let prev_end = prev
+                    .subtree_end_height
+                    .expect("End height is known for the second-to-last shard");
+                let block_delta = prev_end - prev.subtree_start_height;
+                let rate = f64::from(0x1 << SAPLING_SHARD_HEIGHT) / f64::from(block_delta);
+                let tip_notes = (f64::from(tip_delta) * rate) as u64;
```

Comment:
This rounds toward zero.
```suggestion
                let tip_notes = (f64::from(tip_delta) * rate).round() as u64;
```

Reply:
This applies to https://github.com/zcash/librustzcash/pull/939